scalar_udf.go

```diff
@@ -0,0 +1,247 @@
+	err := C.CString(msg)
+	C.duckdb_scalar_function_set_error(function_info, err)
+	C.duckdb_free(unsafe.Pointer(err))
+}
+
+//export scalar_udf_callback
+func scalar_udf_callback(function_info C.duckdb_function_info, input C.duckdb_data_chunk, output C.duckdb_vector) {
+	extraInfo := C.duckdb_scalar_function_get_extra_info(function_info)
+
+	// extraInfo is a void* pointer to our ScalarFunc.
+	h := *(*cgo.Handle)(unsafe.Pointer(extraInfo))
+	function := h.Value().(ScalarFunc)
+
+	// Initialize the input chunk.
+	var inputChunk DataChunk
+	if err := inputChunk.initFromDuckDataChunk(input, false); err != nil {
+		setFuncError(function_info, getError(errAPI, err).Error())
+		return
+	}
+
+	// Initialize the output chunk.
+	var outputChunk DataChunk
+	if err := outputChunk.initFromDuckVector(output, true); err != nil {
+		setFuncError(function_info, getError(errAPI, err).Error())
+		return
+	}
+
+	// Execute the user-defined scalar function for each row.
+	executor := function.Executor()
+	values := make([]driver.Value, len(inputChunk.columns))
+	columnCount := len(values)
+
+	var err error
+	for rowIdx := 0; rowIdx < inputChunk.GetSize(); rowIdx++ {
+		// Set the values for each row.
+		for colIdx := 0; colIdx < columnCount; colIdx++ {
+			if values[colIdx], err = inputChunk.GetValue(colIdx, rowIdx); err != nil {
+				setFuncError(function_info, getError(errAPI, err).Error())
+				return
+			}
+		}
+
+		// Execute the function and write the result to the output vector.
+		var val any
+		if val, err = executor.RowExecutor(values); err != nil {
+			break
+		}
+		if err = outputChunk.SetValue(0, rowIdx, val); err != nil {
+			break
+		}
+	}
+
+	if err != nil {
+		setFuncError(function_info, getError(errAPI, err).Error())
+	}
+}
+
+//export scalar_udf_delete_callback
+func scalar_udf_delete_callback(extraInfo unsafe.Pointer) {
+	h := (*cgo.Handle)(extraInfo)
+	h.Delete()
+}
+
+func registerInputParams(config ScalarFuncConfig, f C.duckdb_scalar_function) error {
+	// Set variadic input parameters.
+	if config.VariadicTypeInfo != nil {
+		t := (*config.VariadicTypeInfo).logicalType()
+		C.duckdb_scalar_function_set_varargs(f, t)
+		C.duckdb_destroy_logical_type(&t)
+		return nil
+	}
+
+	// Set normal input parameters.
+	if config.InputTypeInfos == nil {
+		return errScalarUDFNilInputTypes
+	}
+	if len(config.InputTypeInfos) == 0 {
+		return errScalarUDFEmptyInputTypes
+	}
+
+	for i, info := range config.InputTypeInfos {
+		if info == nil {
+			return addIndexToError(errScalarUDFInputTypeIsNil, i)
+		}
+		t := info.logicalType()
+		C.duckdb_scalar_function_add_parameter(f, t)
+		C.duckdb_destroy_logical_type(&t)
+	}
+	return nil
+}
+
+func registerResultParams(config ScalarFuncConfig, f C.duckdb_scalar_function) error {
+	if config.ResultTypeInfo == nil {
+		return errScalarUDFResultTypeIsNil
+	}
+	if config.ResultTypeInfo.InternalType() == TYPE_ANY {
+		return errScalarUDFResultTypeIsANY
+	}
```

Comment:
Is this because we do not support the ANY type? In that case I think the message should mention that we don't support it, since it seams reasonable to support this in the future

Reply:
Ah, that is really odd to not support the any type, but I guess if duckdb doesnt support it we won't have to either :)

Reply:
In DuckDB internally, there are functions with the `ANY` return type. However, we must look at the input parameters during binding and resolve the `ANY` return type (in a custom bind function/callback). This callback is currently missing in the API on the duckdb side. Maybe in the future, if we add support in DuckDB, we can extend the `ScalarFuncExecutor` with the corresponding `ResolveReturnType` function type.

In `scalar_function.hpp`: 
```cpp	
//! The bind function (if any)
bind_scalar_function_t bind;
```